devices/flash-flexspi/fspi/fspi.c

```diff
@@ -0,0 +1,476 @@
+{
+	unsigned int i;
+
+	if (count > 4) {
+		count = 4;
+	}
+
+	for (i = 0; i < count; ++i) {
+		/* Clear AHB access split control and reset chip size */
+		u32 reg = (*(fspi->base + flsha1cr0 + i) & (0xff << 23));
+
+		/* Assign flash size to ports that are initialized */
+		if (fspi->slPortMask & (1 << i)) {
+			fspi->slFlashSz[i] = ((flashSizes[i] >> 10) & ((1 << 22) - 1));
+			reg |= fspi->slFlashSz[i];
+			fspi->slFlashSz[i] <<= 10;
+		}
+
+		*(fspi->base + flsha1cr0 + i) = reg;
+	}
+}
+
+
+__attribute__((section(".noxip"))) static addr_t flexspi_getAddressByPort(flexspi_t *fspi, u8 port, addr_t addr)
+{
+	unsigned int i;
+
+	/* FlexSPI use the port (chip select) based on an offset of each memory size */
+	for (i = 0; i < port; ++i) {
+		if (fspi->slPortMask & (1 << i)) {
+			addr += fspi->slFlashSz[i];
+		}
+	}
+
+	return addr;
+}
+
+
+int flexspi_deinit(flexspi_t *fspi)
+{
+	/* Leaving initialized */
+
+	return EOK;
+}
+
+
+__attribute__((section(".noxip"))) static int flexspi_checkFlags(flexspi_t *fspi)
+{
+	u32 flags = *(fspi->base + intr) & ((1 << 1) | (1 << 3) | (1 << 11));
+
+	if (flags != 0) {
+		/* Clear flags */
+		*(fspi->base + intr) |= flags;
+
+		/* Reset FIFOs */
+		*(fspi->base + iptxfcr) |= 1;
+		*(fspi->base + iprxfcr) |= 1;
+
+		/* Command grant or sequence execution timeout */
+		if (flags & ((1 << 11) | (1 << 1))) {
+			return -ETIME;
+		}
+
+		return -EIO;
+	}
+
+	return EOK;
+}
+
+
+__attribute__((section(".noxip"))) static ssize_t flexspi_opRead(flexspi_t *fspi, time_t start, struct xferOp *xfer)
+{
+	int res;
+	size_t ofs, len = xfer->data.read.sz, size = xfer->data.read.sz & ~7;
+	u8 *buf = xfer->data.read.ptr;
+
+	/* note: FlexSPI FIFO watermark level is 64bit aligned */
+	for (ofs = 0; ofs < size; ofs += 8, buf += 8, len -= 8) {
+		/* Wait for rx FIFO available */
+		while ((*(fspi->base + intr) & (1 << 5)) == 0) {
+			res = flexspi_checkFlags(fspi);
+			if (res != EOK) {
+				return res;
+			}
+			else if (xfer->timeout > 0 && (hal_timerGet() - start) >= xfer->timeout) {
+				return -ETIME;
+			}
+		}
+
+		((u32 *)buf)[0] = (fspi->base + rfdr32)[0];
+		((u32 *)buf)[1] = (fspi->base + rfdr32)[1];
+
+		/* Move FIFO pointer to watermark level */
+		*(fspi->base + intr) |= 1 << 5;
+	}
+
+	if (ofs < xfer->data.read.sz) {
+		/* Wait for rx FIFO available */
+		while ((*(fspi->base + intr) & (1 << 5)) == 0) {
+			res = flexspi_checkFlags(fspi);
+			if (res != EOK) {
+				return res;
+			}
+			else if (xfer->timeout > 0 && (hal_timerGet() - start) >= xfer->timeout) {
+				return -ETIME;
+			}
+		}
+
+		for (ofs = rfdr32; len > 0; ++ofs) {
+			u32 tmp = *(fspi->base + ofs); /* is volatile! */
+			size = len < sizeof(tmp) ? len : sizeof(tmp);
+			hal_memcpy(buf, &tmp, size);
+			len -= size;
+			buf += size;
+		}
+		/* Move FIFO pointer to watermark level */
+		*(fspi->base + intr) |= 1 << 5;
+	}
+
+	/* FIXME: delay of ~27us */
+	for (u32 i = 0x10000; i > 0; --i) {
+		asm volatile("nop");
+	}
```

Comment:
While using `hal_timerGet()` instead?

Reply:
I've already tried this, `hal_timerGet()` counts in milliseconds, and 1ms is too long should be around 27us.

Reply:
Ok, so maybe we need higher precision timer. FIXME for later